CB-8043 CB-6462 CB-6105 Refactor orientation preference support

MMrko opened the pull request:
Fixes [CB-8043](https://issues.apache.org/jira/browse/CB-8043), [CB-6462](https://issues.apache.org/jira/browse/CB-6462) and [CB-6105](https://issues.apache.org/jira/browse/CB-6105). Touches [CB-6182](https://issues.apache.org/jira/browse/CB-6182).

Adds support for the global & platform-specific orientation preference (iOS, WP8 and FirefoxOS). Android and Amazon Fire OS already had this functionality. Additionally, the change will allow specifying a custom value  for the platform-specific orientation preference.  If need be I'd be happy to update the [docs](http://cordova.apache.org/docs/en/4.0.0/config_ref_index.md.html) accordingly.

As a future improvement the orientation preference could offer a similar level of granularity to that of the native APIs. For example, iOS & WP8 APIs allow setting both an initial orientation as well as a set of supported orientations for an app. For now, developers will have to resort to a custom approach in their Cordova builds in case they've such needs.

Tested on OS X Mavericks & Windows 8 running VS 2013.

MMrko commented on cordova-lib/src/configparser/ConfigParser.js:
```diff
@@ -219,6 +258,47 @@ ConfigParser.prototype = {
+        platformOrientation = this.getPlatformPreference('orientation', platform);
+
+        // Check if the value of the platform-specific orientation preference is invalid
+        if (platformOrientation && !_.contains(platformSupportedOrientations, platformOrientation)) {
```
Yep, that would've been a better approach. In the end supportedOrientations.json seemed to be useful only for validating user input so I omitted this check for platform-specific support altogether.

MMrko commented:
Thanks for the feedback! 

1. True, pardon my ignorance :) I removed the style-only changes
2. I moved the new functionality in to a separate [helper module](/MMrko/cordova-lib/tree/fix-orientation-pref/cordova-lib/src/cordova/metadata/helpers/). The platform parsers now use this helper module by inheriting a [base parser](/MMrko/cordova-lib/tree/fix-orientation-pref/cordova-lib/src/cordova/metadata/parser.js), which the helper module extends

The platform parsers required the most refactoring. Previously the methods in each platform parser were assigned in a way that overrode the prototype. This needed to be refactored to allow the platform parsers to inherit from a base parser without overriding it (see 918b47eba0308869b38a91a83055c26c15791bb6).

agrieve commented:
A few comments but love where this is going! I've merged in your first 3 commits since those are fine and I don't want to risk merge conflicts by leaving them pending for a while.

MMrko commented:
Validation is now done only on the global orientation preference. Once this is good for merge let me know and I'll rebase for a cleaner history.

MMrko commented:
Rebased on master (0f5dbafface487c5738f294e47520d8b2381134d). I squashed the changes into 3 commits:

6ef506ebd890bf3eca79f8bb034ee9e2ede1c948 Base parser & helper related changes
2652a190457d79f988d3f6e5f678e490d7a55b39 Fixes to `firefoxos_parser.js`
179a3b2cf256890b163c2f3f7a63a9b41dcc09b8 Changes to the platform parsers (closes #128)

MMrko commented:
Thanks a lot @agrieve for taking the time to review all of this! Really helpful :)

agrieve commented:
No problem! Thanks for being receptive and for taking the time to contribute!

VincentFTS commented:
Hello,

I'm using cordova 4.3.0 and I can't get orientation lock working on ipad. What am I missing ?

mmrko commented:
Thanks for reporting this @VincentFTS. Doesn't seem to work for me either using the iOS Simulator. Have you tested this on a device?

VincentFTS commented:
Yes I do, same issue.

I've just seen that in the plist file, man can find the UISupportedInterfaceOrientations~ipad key which correspond to the UISupportedInterfaceOrientations for iPad. Maybe cordova should modify both.

mmrko commented:
Okay, makes sense. Thanks! I'll look into this over the weekend. Or, if you find the time feel free to put up a PR for this :)

rachel-carvalho commented:
This iPad issue is affecting my project, so I took the liberty of trying to implement it, I'll be sending a PR.
